README

			       asmutils
			       ********

			    (version 0.18)

			http://linuxassembly.org


Legal stuff
===========

 All programs in this package are *copyrighted* by their authors,
 as stated in the source code and in the CREDITS file. The bulk
 of code is written by Konstantin Boldyshev, with contributions
 of various hackers all over the world.

 These programs are free software; you can redistribute and/or modify
 them under the terms of the version 2 (and only version 2) of the GNU
 General Public License as published by the Free Software Foundation.

 These programs are distributed in the hope that they will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty
 of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the version 2 of the GNU General Public License for more details
 <http://www.gnu.org/copyleft/gpl.html>.

 Authors are not responsible for any damage these programs can do.
 This is development version, use solely at your own risk.

Description
===========

 asmutils is a set of miscellaneous utilities written in assembly language,
 targeted on embedded systems and small distributions (e.g. installation
 or rescue disks); also it contains a small libc and a crypto library.
 It features the smallest possible size and memory requirements,
 the fastest speed, and offers fairly good functionality.
 Now project supports Linux and BSD family (FreeBSD, OpenBSD, NetBSD);
 Unixware, Solaris, and AtheOS support is in beta stage (also there's
 no more supported BeOS port), and it is /theoretically/ possible to
 port it to any i386 OS. 

 Another goal of this package is to provide a *portable* framework,
 and to encourage assembly programmers to write for Linux/UNIX.
 Usually UNIX is associated with C, and no one wants to deal with assembly..
 Browse the source, look -- it is simple. Read the docs, start to write,
 and you will like it; assembly programming is a lot of fun, especially
 when you deal with a UNIX-like operating systems.

 In addition, this package has some "special bonus" like small intros,
 games, and other various assembly snippets.

 These utilities also may be useful for system administrators;
 there were several reports that they have rescued when libc
 crashed and it was not possible to use usual system utilities.

 Hey, we are a light in the darkness of bloatware. Do you see?

How to install
==============

 REQUIREMENTS:

	1) i386 (however some programs may require higher CPU);
	   some programs may require FPU (or kernel emulation)

	2) UNIX-like kernel

	3) ELF/AOUT executable format (actually not a requirement,
	   you just need to write startup stack preparation
	   for your executable format on your own)

 For details on compilation process examine doc/Asmutils-HOWTO,
 then edit MCONFIG, run '[g]make' and hopefully all will be done.

 '[g]make install' creates bin directory in the source tree and stores
 compiled files there (it doesn't copy files elsewhere, neither to /bin
 or /usr/bin, nor to /usr/local/bin).

 You may also want to examine a-Linux -- a floppy-based mini-distribution
 consisting of Linux kernel and asmutils; it is available for download
 from the asmutils homepage as well.

 WARNING:
 
 Before using any utility first DO LOOK AT ITS SYNTAX; some utils may
 work not as you have expected. Syntax can be found in the appropriate
 source file (.asm) from the src directory (do not expect any answer
 from program on --help switch or without switches).


Feedback
========

 Contributions are welcome! If you want to contribute some code,
 please read the Asmutils-HOWTO in the doc directory.
 At least send a bugreport or a success story :)

 $Id: README,v 1.17 2006/02/18 10:13:42 konst Exp $


ChangeLog
18-Feb-2006:	asmutils 0.18

	- cda2raw, idea, md5 (MH)
	- ksyms, lsusb (NK)
	- deflate/inflate (SW)
	- telnetd (RM)
	- users (JC)
	- gi (JH)
	- tar: bugfixes, archive creation support (JH)
	- dd: fixed pipe bug, added conv=sync conv=swab,
	  count blocks, recover lseek (JH)
	- httpd: no longer runs as root if UID defined (JH),
	  sendfile support (RM),
	  fixed 404 error messages for paths with %20's when PROC_HANDLE is defined (RHS)
	- sh: tab completion fixes (RM),
	  improved jobs, added umask (JH),
	  added help, enable, pushd, popd, dirs internals (NK)
	- cat: added short form of all GNU options (JH)
	- du: added "-d" option (JH)
	- host: added "-t" option (WT)
	- chmod: fixed "chmod 4755" bug (KB)
	- getty, finger: fixed "lea" compile error (KB)
	- special bonus: fbcon intro hdragon.asm (MH)
	- fire256: misc improvements (PF)
	- system.inc: "last writable section" fix for Linux 2.6.11,
	  added CPU 386 directive
	  (also added higher CPU directives to few utilities) (KB)
	- moved bonus stuff to src/bonus (KB)

17-Aug-2002:	asmutils 0.17

	- free, getty, scons, truss, uptime, wget (TO)
	- cal, netstat (SL)
	- du, mkfs.minix (JH)
	- ftpd (AM, RM)
	- proxy (IM)
	- uuencode (CB)
	- sha1sum (JD)
	- nm: added 'size' extension (TO, KB)
	- finger: added 'fingerd' extenstion (TO)
	- swapon: added 'mkswap' extension (TO)
	- sleep: added 'usleep' extension (IM)
	- sh: added clear internal (KB)
	- httpd: added initial cgi support (SL),
	  '%' support in filenames (RM),
	  send default mimetype for unknown extension (KB)
	- less: added dynamic window size, stdin redirection check,
	  new keyboard keys support, optimization (KB)
	- finger: added individual user lookup (TO)
	- pwd: bugfixes, syscall optimization, and portability fixes
	  in stat-based version (KB)
	- cmp: bugfixes and size improvements (KB)
	- yes: size improvements (KB)
	- ifconfig: exit code fixes (KB)
	- mount: try to mount ro when rw fails (KB)
	- init: wait for rc script to finish (KB)
	- tar: added contiguous (append) files,
	  prefix processing,  "tar -xf -" for stdin, chown/grp,
	  selection of only certain filenames for "tar -x" (JH)
	- tail: added support for -n CHARSc (JH)
	- tr: fixed null character processing (JH)
	- libc: fixed __start_main/fastcall bug (KB)
	- special bonus: fbcon intro mandelbrot.asm (SW)

03-Mar-2002:	asmutils 0.16

	- cmp, finger, killall/killall5/pidof, mesg, nice, nm, nohup,
	  rot13, sysctl, test, write (TO)
	- fdisk, host, tar, telnet (RM)
	- cut, m_inetd (JH)
	- pop3d (MM)
	- scan (SR)
	- kldload/kldunload/kldstat (KB)
	- sh: added && and || internals and scripting (JH),
	  added pipe & redirection support, wildcard extending,
	  ctrl+z handling, fg, bg, jobs internal commands (RM),
	  added $environment variable handling and some control-characters (TO),
	  '#' comments, improved scripting, cleanup and misc fixes (KB)
	- httpd: added logging (IP||HEADER), err404file command line argument,
	  added more content types (RM), added extension-content type table,
	  fixed infinite loop when err404file is missing, size improvement (KB)
	- grep: added "-b", "-c", "-v" options,
	  output filename when grepping several files (KB)
	- hostname: added sysctl-based version (KB)
	- uname: fixed sysctl-based BSD version (KB)
	- mount: rewritten to match usual syntax, BSD port (KB)
	- reboot: added "-p", "-f" options, safety check, BSD port (KB)
	- lsmod: use sys_query_module if /proc/modules is missing (KB)
	- ls: fixed FPE when file's mtime is in the future (KB)
	- touch: fixes and size improvement (KB)
	- head: fixed reading from stdin (JH)
	- init: execute rc script on startup (KB)
	- libc: fixed strlen(), other various fixes (KB)
	- initial Solaris port (TO)
	- initial UnixWare port (TO)
	- rewritten libc port (KB)
	- ELF macros for BSD and perhaps all other OSes (KB)
	- include files bugfixes, uptates, improvements and rearrangement (KB)
	- bootdisk.bash: added creation of /etc/rc startup script (KB)
	- Asmutils-HOWTO: misc updates (KB)

24-Sep-2001:	asmutils 0.15

	- dc, mknod/mkfifo, od, paste, readlink, tr, which (JH)
	- clear, fromdos/todos, less, readkey, whoami (TG)
	- ifconfig/route, tail (JM)
	- readelf (RR)
	- watch (RM)
	- sh: added char-oriented commandline, tab-filename filling,
	  partial export support, partial CTRL+C handling,
	  added history handling, improved signal handling (RM)
	- dd: fixed "k" suffix bug (RM)
	- true: even more size improvements (RM)
	- id: added groups=list (JH)
	- wc: fixed option handling without arguments (JH)
	- touch: fixed compilation bug (KB)
	- various FreeBSD and LIBC build fixes (KB)
	- added bootdisk.bash script to build boot-disk image
	  with Linux kernel and asmutils binaries (KS)
        - libm: fixed minor bugs, improved optimizations and test suite,
	  support for different IA32-CPUs (mostly i686) (NK)
        - heap.asm: heap manager for Linux (RM)
	- special bonus: battle "game", resembling corewars (RM)

18-Mar-2001:	asmutils 0.14

	- date, hexdump, ls (BR)
	- init (KS)
	- dd (RM)
	- ping (JM)
	- factor: increased upper limit to 2^64-1, size improvement (BR)
	- mount: support for mount options (RM)
	- libc: added getenv(), optionally treat stack with care (KB)
	- libm: added testm (NK)
	- special bonus: textmode game snake (BR)
	- Asmutils-HOWTO: misc updates (KB)


23-Feb-2001:	asmutils 0.13

	- sleep: added nanoseconds support (KB)
	- libc: added startup stub, PIC fixes (KB)
	- libm: various improvements (NK)
	- various FreeBSD build fixes (KB)
	- include file updates (KB)


21-Jan-2001:	asmutils 0.12

	- chown, renice, touch (JL)
	- asmutils multicall binary (KB)
	- httpd: support for "Content-Type: text/plain" for text files (KB)
	- lzss: bugfixes (KB)
	- libc: rewritten and refined syscall mechanism,
	  usual functions now work with both cdecl and fastcall,
	  added PIC support, significant size improvement,
	  lot of other fixes (KB), added libm (NK)
	- system.inc: added _xchg macro (PR), _jmp macro (KB)

src/dd.asm
;Copyright (C) 2001 Rudolf Marek <marekr2@feld.cvut.cz>
;
;$Id: dd.asm,v 1.5 2003/06/20 18:53:11 konst Exp $
;
;hackers' dd
;
;syntax: dd if= of= count= skip= bs= seek= conv=(sync|swab)
; swab requires an even bs
;
;number can be also 1k=1*1024 etc m=1024 * 1024, k=1024 b=512 w=2
;
;0.1: 2001-Feb-21	initial release
;0.2: 2002-Apr-15	added O_LARGEFILE for input file
;0.3: 2003-Jun-18	fixed pipe bug, added conv=sync conv=swab
;			count blocks, recover lseek (JH)
;
;All comments/feedback welcome.

%include "system.inc"

%define SIMPLE_CONV
%define LSEEK_RECOVER
%define COUNT_BLOCKS
%define ERROR_MSG

%ifdef	__LINUX__
%define LARGE_FILES
%endif

CODESEG

usage	db	"usage: dd as you know except ibs= obs= conv=",__n
_ul	equ	$-usage
%assign	usagelen _ul
%ifdef ERROR_MSG
errmsg	db	"IO-error", 10
errlen	equ	$-errmsg
%endif

START:  
	_mov 	ebp,STDOUT
	_mov 	edi,STDIN 
	pop     eax                     ;argc
	dec     eax
	pop     eax                     ;argv[0], program name
	jnz      .continue

	sys_write ebp,usage,usagelen
	sys_exit 0

.continue:
	
	mov	byte [bs+1],0x2		;bs = 512 - default block size
	
.next_arg:
	pop	esi
	or 	esi,esi
	jz 	near .no_next_arg

.we_have_arg:
	push 	dword .next_arg

	cmp 	word [esi],'of'
	jz 	.parse_output_file
	cmp 	word [esi],'if'
	jz 	.parse_input_file
	mov	edx,count
	cmp 	dword [esi],'coun'
	jz 	.update_fields	
	add	edx,byte 4
	cmp 	word [esi],'bs'
	jz 	.update_fields	
	add	edx,byte 4
	cmp 	dword [esi],'skip'
	jz 	.update_fields	
	add	edx,byte 4
	cmp 	dword [esi],'seek'
	jz 	.update_fields	
%ifdef SIMPLE_CONV
	cmp	dword [esi],'conv'
	je	.conv
%endif

	ret  ;ignore unknown opt

.parse_input_file:		;I always wanted to xchange .. :)
	_mov	ecx,(O_RDONLY|O_LARGEFILE)
	xchg	edx,ebp
	call	.open
	xchg	edi,ebp
	xchg	edx,ebp
	ret

.parse_output_file:	; No no, not O_TRUNC. Try to improve GNU dd, not copy
	_mov	ecx,(O_WRONLY|O_CREAT|O_LARGEFILE)
	_mov	edx,(S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH) ;dd feeds _always_ 666
.open:	
	call 	.check
	xchg 	ebx,esi
	sys_open
	test 	eax,eax
	js 	.do_error
	xchg 	ebp,eax
	ret

.update_fields:
	call 	.check
	call 	.ascii_to_num
	mov 	[edx],eax
	ret
%ifdef SIMPLE_CONV
.conv:		; Turn on rudimentry conversion (sync & swab)
	add	esi, byte 5
	cmp	[esi], dword 'sync'
	je	.c_sync
	cmp	[esi], dword 'swab'
	je	.c_swab
	ret
%endif

.check:
	lodsb
	or 	al,al
	jz	.error
.ok:
	cmp 	al,'='
	jnz 	.check
	ret
.do_error:
%ifdef ERROR_MSG
	sys_write	2, errmsg, errlen
%endif
%ifdef COUNT_BLOCKS
	call	.showcount
%endif
.error:
	sys_exit 1

%ifdef SIMPLE_CONV
.c_sync	mov	[c_sync], byte 1
	ret
.c_swab	mov	[c_swab], byte 1
	ret
%endif

.no_next_arg:			;now we should have opened files - ready to copy
	mov 	eax,[bs]
	push	eax
	add	eax,buf
	sys_brk eax            ;get some mem
	pop	eax
	mul 	dword [skip]   ;EDX:EAX seek input file

%ifdef LARGE_FILES
	push    edi
	mov     ebx,edi
	mov     ecx,edx
	mov     edx,eax
	_mov     esi,result
	_mov     edi,SEEK_SET
	sys_llseek
	;pop	edi
%else	
	or 	edx,edx		;file bigger than 4Gb cannot lseek more
%ifdef LSEEK_RECOVER		;so skip blocks if compiled in
	jnz	.recover
%else
	jnz 	.do_error	
%endif
	sys_lseek edi, eax, SEEK_SET
%endif

%ifdef LSEEK_RECOVER
	or	eax, eax	; Trick here is to read [skip] full blocks
	jns	.norecover
.recover:
	mov	ebx, esi
	mov	ecx, buf
.rec_nextblock:
	dec	dword [skip]
	js	.norecover	; Already skipped enough?
	mov	edx, [bs]
.rec_nextread:
	sys_read
	or	eax, eax
	jna	.do_error
	sub	edx, eax
	jnz	.rec_nextread
	jmp	short	.rec_nextblock
.norecover:
%endif

	mov 	eax,[bs]	; Now seek nicely on output file
	mul 	dword [seek]
%ifdef LARGE_FILES
	mov     ebx,ebp
	mov     ecx,edx
	mov     edx,eax
	;push    edi
	;mov     esi,result
	;mov     edi,SEEK_SET
	sys_llseek
	pop	edi
%else
	or 	edx,edx		; Can't skip here
	jnz 	.do_error	
	sys_lseek ebp,eax,SEEK_SET
%endif
	mov 	esi,[count]
.next_block:
	; The idea here is to keep reading until [bs] bytes have been read
	mov	ebx, edi
	_mov	ecx, buf
	_mov	edx, [bs]
.next_read:
	sys_read
	test 	eax,eax
	jz 	.no_more_data
	js 	near .do_error
	add	ecx, eax
	sub	edx, eax
	ja	.next_read
%ifdef COUNT_BLOCKS
	inc	dword [incnt]
%endif

%ifdef SIMPLE_CONV
	call	.swab
%endif

	mov 	ebx,ebp
	mov 	edx,ecx
	mov	ecx,buf
	sub	edx,ecx
	sys_write
%ifdef ERROR_MSG
	cmp	eax, edx
	jne	near .do_error
%endif
%ifdef COUNT_BLOCKS
	inc	dword [outcnt]
%endif
	dec 	esi
	jnz	.next_block
	xor	eax, eax
.no_more_data:	
%ifdef COUNT_BLOCKS
	cmp	edx, [bs]
	je	.finalize	; Guess what, we can skip all of this
	inc	byte [inpart]
	inc	byte [outpart]
%endif
%ifdef	SIMPLE_CONV
	cmp	[c_sync], byte 1
	jne	.flush
	xor	eax, eax
	mov	edi, ecx
	mov	ecx, edx
	rep	stosb	; Actually loops 0 times on zero
	mov	ecx, edi
%ifdef COUNT_BLOCKS
	dec	byte [outpart]
	inc	byte [outcnt]
%endif
.flush: 
	call	.swab
%endif
	mov 	edx,ecx
	mov	ecx,buf
	sub	edx,ecx
	sys_write ebp
.finalize:
	sys_close edi
	sys_close ebp
%ifdef	COUNT_BLOCKS
	call	.showcount
%endif
	sys_exit 0	

;Swab buffer code
; ecx = end of buffer
; all registers conserved
.swab:	test	[c_swab], byte 1
	jz	.ret
	pusha
	mov	esi, buf
	mov	edi, esi
	sub	ecx, esi
	or	ecx, ecx
	jz	.trap
.swabber:
	lodsw
	xchg	al, ah
	stosw
	loop	.swabber
.trap:	popa
.ret	ret

;Here is long output count code-------------
%ifdef COUNT_BLOCKS
.showcount:
	mov	edi, buf
	push	edi
	mov	eax, [incnt]
	call	.itoa
	xor	eax, eax
	mov	ah, [inpart]
	add	eax, "+0 "
	stosd
	mov	eax, 0x0A2C6E69		; "in,", __n
	stosd
	mov	eax, [outcnt]
	call	.itoa
	xor	eax, eax
	mov	ah, [outpart]
	add	eax, "+0 "
	stosd
	mov	eax, 0x0A74756F		; "out,", __n
	stosd
	pop	ecx
	mov	edx, edi
	sub	edx, ecx
	sys_write	STDERR
	ret

.itoa:
	_mov	ebx, 10
	xor	ecx, ecx
.itoan	xor	edx, edx
	div	ebx
	add	dl, '0'
	push	edx
	inc	ecx
	or	eax, eax
	jnz	.itoan
.itoap	pop	eax
	stosb
	loop	.itoap
	ret
%endif

;---------------------------------------stolen from renice.asm
; esi = string
; eax = number 
.ascii_to_num:
	;push	esi
	xor     eax,eax                 ; zero out regs
	xor     ebx,ebx
	
	;cmp     [esi], byte '-'
	;jnz     .next_digit
	;lodsb

.next_digit:
	lodsb                           ; load byte from esi
	or	al,al
	jz	.done
	cmp 	al,'9'
	ja  	.multiply  
	sub	al,'0'                  ; '0' is first number in ascii
	imul	ebx,10
	add	ebx,eax
	jmp	short	.next_digit

.done:
	xchg    ebx,eax
	;pop	esi
	;cmp     [esi], byte '-'
	;jz     .done_neg
	ret
;.done_neg:
;	neg	eax			;if first char is -, negate
;	ret
.multiply:
	cmp 	al,'w'
	jz	 .mul_2
	cmp 	al,'b'
	jz 	.mul_512
	cmp	al, 'k'
	jz	.mul_1024
	cmp 	al,'m'
	jnz	near .error		;we don't know others yet

	shl	ebx,10
.mul_1024:
	shl	ebx,1
.mul_512: 
	shl	ebx,8
.mul_2:
	shl	ebx,1
	jmp	short .done
		
;---------------------------------------	

UDATASEG

count	resd	1
bs	resd	1
skip	resd	1
seek	resd	1

incnt	resd	1
outcnt	resd	1
inpart	resb	1
outpart	resb	1

c_sync	resb	1
c_swab	resb	1

result	resq	1
buf	resb	1	;here will our buff start

END
